graphql_compiler/schema_transformation/split_query.py

```diff
@@ -39,6 +47,165 @@ def __init__(self, query_ast):
+          selection, or some inline fragment coexists with other fields or inline fragment
+          on the same scope
+        - SchemaStructureError if the input merged_schema_descriptor appears to be invalid
+          or inconsistent
+    """
+    check_query_is_valid_to_split(merged_schema_descriptor.schema, query_ast)
+
+    # If schema directives are correctly represented in the schema object, type_info is all
+    # that's needed to detect and address stitching fields. However, GraphQL currently ignores
+    # schema directives when converting an AST to a schema object. Until this issue is
+    # fixed, it's necessary to use additional information from pre-processing the schema AST
+    edge_to_stitch_fields = _get_edge_to_stitch_fields(merged_schema_descriptor)
+    name_assigner = IntermediateOutNameAssigner()
+
+    root_query_node = SubQueryNode(query_ast)
+    query_nodes_to_split = [root_query_node]
+
+    # Construct full tree of SubQueryNodes in a dfs pattern
+    while len(query_nodes_to_split) > 0:
+        current_node_to_split = query_nodes_to_split.pop()
+
+        _split_query_one_level(current_node_to_split, merged_schema_descriptor,
+                               edge_to_stitch_fields, name_assigner)
+
+        query_nodes_to_split.extend(
+            child_query_connection.sink_query_node
+            for child_query_connection in current_node_to_split.child_query_connections
+        )
+
+    return root_query_node, frozenset(name_assigner.intermediate_output_names)
+
+
+def _get_edge_to_stitch_fields(merged_schema_descriptor):
+    """Get a map from type/field of each cross schema edge, to the fields that the edge stitches.
+
+    This is necessary only because GraphQL currently doesn't process schema directives correctly.
+    Once schema directives are correctly added to GraphQLSchema objects, this part may be
+    removed as directives on a schema field can be directly accessed.
+
+    Args:
+        merged_schema_descriptor: MergedSchemaDescriptor namedtuple, containing a schema AST
+                                  and a map from names of types to their schema ids
+
+    Returns:
+        Dict[Tuple(str, str), Tuple(str, str)], mapping (type name, vertex field name) to
+        (source field name, sink field name) used in the @stitch directive, for each cross
+        schema edge
+    """
+    edge_to_stitch_fields = {}
+    for type_definition in merged_schema_descriptor.schema_ast.definitions:
+        if isinstance(type_definition, (
+            ObjectTypeDefinition, InterfaceTypeDefinition
+        )):
+            for field_definition in type_definition.fields:
+                stitch_directive = try_get_ast_by_name_and_type(
+                    field_definition.directives, u'stitch', Directive
+                )
+                if stitch_directive is not None:
+                    source_field_name = stitch_directive.arguments[0].value.value
```

Comment:
This is actually buggy in a very subtle and non-obvious way: directive arguments are passed by name, not positionally – so I could in principle pass `@stitch(sink_field: "foo", source_field: "bar")` and cause this code to load them backwards.

The `compiler_frontend.py` file works around this by passing `directive.arguments` to `get_uniquely_named_objects_by_name` and then loading argument values by name. I'd suggest that here as well.

Reply:
Ah good point. The current iteration of `merge_schemas` always creates `@stitch` directives in the order of source to sink, but you're definitely right that I should check arguments by name to be safe.

On that note, should I make a test where the `@stitch` arguments are switched? It's a little awkward to make a schema for that test, since I'll need to create a MergedSchemaDescriptor by hand instead of by `merge_schemas`

Reply:
Probably a good idea to have a test, even if it requires making a descriptor by hand. It's the sort of thing where (now that we've spotted the potential issue) we don't expect problems in the future, but if a future refactor caused a regression then we'd lose a lot of time in debugging. Better to invest a bit of time now in a good test instead.